File: packages/core/src/factories/internal/metadata/iterate_metadata_set.ts
import { ArrayUtil } from "@typia/utils";
import ts from "typescript";

import { MetadataSchema } from "../../../schemas/metadata/MetadataSchema";
import { MetadataSet } from "../../../schemas/metadata/MetadataSet";
import { TypeFactory } from "../../TypeFactory";
import { IMetadataIteratorProps } from "./IMetadataIteratorProps";
import { explore_metadata } from "./explore_metadata";

export const iterate_metadata_set = (
  props: IMetadataIteratorProps,
): boolean => {
  const type: ts.Type = props.checker.getApparentType(props.type);

  const name = TypeFactory.getFullName({
    checker: props.checker,
    type: type,
    symbol: type.getSymbol(),
    aliasTypeArguments: false,
  });
  const generic = props.checker.getTypeArguments(type as ts.TypeReference);
  if (name.substring(0, 4) !== "Set<" || generic?.length !== 1) return false;

  const key: ts.Type = generic[0]!;
  const value: MetadataSchema = explore_metadata({
    ...props,
    type: key,
    explore: {
      ...props.explore,
      escaped: false,
      aliased: false,
    },
    intersected: false,
  });
  ArrayUtil.take(
    props.metadata.sets,
    (elem) => elem.value.getName() === value.getName(),
    () =>
      MetadataSet.create({
        value: explore_metadata({
          ...props,
          type: key,
          explore: {
            ...props.explore,
            escaped: false,
            aliased: false,
          },
          intersected: false,
        }),
        tags: [],
      }),
  );
  return true;
};

File: packages/core/src/factories/internal/metadata/iterate_metadata_set.go
package metadata

func IterateMetadataSet(props IMetadataIteratorProps) bool {
	if props.Type == nil || !props.Type.IsSet() || len(props.Type.TypeArguments()) != 1 {
		return false
	}
	keyType := props.Type.TypeArguments()[0]
	value := ExploreMetadata(ExploreMetadataProps{Options: props.Options, Checker: props.Checker, Components: props.Components, Errors: props.Errors, Type: keyType, Explore: props.Explore.WithAliased(false).WithEscaped(false)})
	addSet(&props.Metadata.Sets, MetadataSet{Value: value})
	return true
}
